Sources/GRPC/GRPCNIO.swift

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import NIOTransportServices
+
+/// How a network implementation should be chosen.
+public enum NetworkPreference {
+  /// Use the best available, that is, Network.framework (and NIOTransportServices) when it is
+  /// available on Darwin platforms (macOS 10.14+, iOS 12.0+, tvOS 12.0+, watchOS 6.0+), and
+  /// falling back to the POSIX network model otherwise.
+  case best
+
+  /// Use the given implementation. Doing so may require additional availability checks depending
+  /// on the implementation.
+  case userDefined(NetworkImplementation)
+}
+
+/// The network implementation to use: POSIX sockets or Network.framework. This also determines
+/// which variant of NIO to use; NIO or NIOTransportServices, respectively.
+public enum NetworkImplementation {
+#if canImport(Network)
+  /// Network.framework (NIOTransportServices).
+  @available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+  case networkFramework
+#endif
+  /// POSIX (NIO).
+  case posix
+}
+
+extension NetworkPreference {
+  /// The network implementation, and by extension the NIO variant which will be used.
+  ///
+  /// Network.framework is available on macOS 10.14+, iOS 12.0+, tvOS 12.0+ and watchOS 6.0+.
+  ///
+  /// This isn't directly useful when implementing code which branches on the network preference
+  /// since that code will still need the appropriate availability check:
+  ///
+  /// - `@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)`, or
+  /// - `#available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)`.
+  public var implementation: NetworkImplementation {
+    switch self {
+    case .best:
+      #if canImport(Network)
+      guard #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *) else {
+        // This is gated by the availability of `.networkFramework` so should never happen.
+        fatalError(".networkFramework is being used on an unsupported platform")
+      }
+      return .networkFramework
+      #else
+      return .posix
+      #endif
+
+    case .userDefined(let implementation):
+      return implementation
+    }
+  }
+}
+
+// MARK: - Generic Bootstraps
+
+/// This protocol is intended as a layer of abstraction over `ClientBootstrap` and
+/// `NIOTSConnectionBootstrap`.
+public protocol GenericClientBootstrap {
```

Comment:
How about e.g. `TransportAgnosticClientBootstrap`? Just an idea; not sure if that's any better.

Also, couldn't this be part of NIO, given its universal usefulness?

Reply:
Hmm I'm not sure about the naming. How about `ClientBootstrapProtocol`? Swift has `StringProtocol` and `IteratorProtocol` so I don't think that it's too much of a stretch.

Reply:
Re: NIO: https://github.com/apple/swift-nio/issues/796

Reply:
Sounds good!

Reply:
But should we add a TODO to remove this once https://github.com/apple/swift-nio/issues/796 is fixed?

Reply:
But should we add a TODO to remove this once https://github.com/apple/swift-nio/issues/796 is fixed?